Adds a new prefab: prefab_saferoom

MetricDuck opened the pull request:
[FEATURE] [MAPPING] [SPRITES]
## About the PR
This is a resubmission of PR #6041 as I messed it up somehow. The inital recommit does not address all outstanding comments on the previous PR.

This prefab is my submission for the 'April' Goonstation Monthly Contest (https://forum.ss13.co/showthread.php?tid=16992).

It's a small, puzzle based prefab with a bit of a story attached (I was going for something along the lines of walking into the aftermath of a BM/TZ episode, YMMV on whether I achieved it!).

Concrete stats:
- Size: 33x22
- Probability: 15
- Hazards: Environmental (mitigatable)
- New loot: Health scan implant that remote clonescans with a chance of adding puritan (See note).
- Other loot & considerations: Some drugs & cash. Also prefab contains most of a pre-built functioning lab/bar which might have balance implications.

Please note I've never done anything in BYOND OR vscode OR github before, and I am still learning as I go! Super grateful for any constructive criticism!

(Note) The implant was mostly for story/gameplay integration, I am more than happy to think of something else if it is verboten!

## Why's this needed?
Contest submission. Also I've always wanted to contribute something to the server and this seemed like a good opportunity.

## Changelog
```changelog
(u)MetricDuck
(*)Added a new space prefab - a seemingly abandoned safe room. 
```

MetricDuck commented:
Comment from Tarmunora on #6041:
_You need to push your modified goonstation.dme, as you're adding a new file_

MetricDuck commented on code/z_adventurezones/safehouse.dm:
```diff
@@ -0,0 +1,718 @@
+		bioHolder.uid_hash = md5(bioHolder.Uid)
+		. = ..()
+
+	New()
+		..()
+
+		SPAWN_DBG(5 SECONDS)
+		for(var/mob/living/carbon/human/dead_exec/M)
+			qdel(M) //Jean's only here to set up the puzzle, we don't want them hanging around.
```
Comment from ZeWaka on #6041 
_this bit of code makes no sense_

MetricDuck replied:
Code removed - no longer used.

MetricDuck commented on code/z_adventurezones/safehouse.dm:
```diff
@@ -0,0 +1,718 @@
+		qdel(src)
+
+	proc/setup_blood_bioHolder()
+		for(var/obj/item/reagent_containers/iv_drip/O in src.loc)
+			if(!O.reagents.has_reagent("blood"))
+				return
+			for(var/mob/living/carbon/human/dead_exec/M in world) //aka Jean
```
Comment from Zewaka on #6041
_why not just spawn the dude inside the object with a variable then delete him_

MetricDuck replied:
You were right that approach was much simpler. Updated accordingly - please see lines 362-378

MetricDuck commented on code/z_adventurezones/safehouse.dm:
```diff
@@ -0,0 +1,686 @@
+				if(7)
+					boutput(H,"ERROR: Remote server unavailable. Commencing emergency transmisssion protocols. Broadcasting data on all known frequencies.")
+				if(10)
+					for(var/obj/machinery/computer/cloning/C in world)
```
Comment from pali on discord
*the for( ... in world) loops are very slow (because they iterate through everything in the world), same with loops of the form for(var/some/type/variable)
a good way to avoid these is to put START_TRACKING in your type's New and STOP_TRACKING in its disposing() and then use
for_by_tcl(variable, /some/type) which will be much faster*

MetricDuck replied:
Replaced main loops with for_by_tcl. Loops that use for(bytype) on poddoors remain, as they're generic poddoors and/or they're using their parent's proc's.

Tarmunora commented on code/z_adventurezones/safehouse.dm:
```diff
@@ -0,0 +1,708 @@
+/obj/machinery/handscanner/bio_handscanner/attackby(obj/item/W, mob/user as mob) //We don't want the parent's proc here.
+	if(istype(W, /obj/item/device/detective_scanner))
+		return
+	if(cooldown == TRUE)
+		return
```
Use `ON_COOLDOWN()` macro for cooldowns

MetricDuck replied:
Cooldowns amended to use macro.

Tarmunora commented on code/z_adventurezones/safehouse.dm:
```diff
@@ -0,0 +1,708 @@
+		STOP_TRACKING
+		..()
+
+/obj/machinery/handscanner/bio_handscanner
```
Is there any reason for this to be a child of handscanner if so much of the behavior is overridden?

MetricDuck replied:
Not really, was just how code was built up as I learned. It’s no longer a child of hand scanner

Tarmunora commented on code/z_adventurezones/safehouse.dm:
```diff
@@ -0,0 +1,708 @@
+							boutput(H,"Link to cloning computer establised succesfully.")
+							playsound(src.loc, 'sound/machines/ping.ogg', 50, 1)
+							if(prob(20) && !H.traitHolder.hasTrait("puritan")) // Too uncommon to weaponise but common enough to deter general usage.
+								R["fields"]["traits"] += "puritan" // Did the player learn nothing from the prefab??
```
Needs to actually add the trait to the traitHolder

MetricDuck replied:
Original implentation applied trait after scan, but bug testing found this ineffective as the clone record didn't include it.

This comment highlighted that reviewed implementation was muddled: it checked current traitholder but applied trait to clone record. This could allow the trait to be applied multiple times through re-implanting.

Amended implementation now checks record and applies trait to clone record (ie: signal degrades, not user). Notable features of this implementation: deleting record and rescanning could fix mistake, doesn't try to apply trait if no cloning records have been generated and overall I think it works more as I would expect it to. 

Rolling for traitHolder.addTrait then attempting to scan_mob (and possibly not succeeding) would be a harsher implementation, but I don't mind amending to that if you think that would be better!

Tarmunora commented on code/z_adventurezones/safehouse.dm:
```diff
@@ -0,0 +1,708 @@
+			if(R)
+				playsound(src.loc, 'sound/machines/ping.ogg', 50, 1)
+				if(prob(20) && !H.traitHolder.hasTrait("puritan")) // Too uncommon to weaponise but common enough to deter general usage.
+					R["fields"]["traits"] += "puritan" // Did the player learn nothing from the prefab??
```
traitholder needs trait

MetricDuck replied:
See other response.

Tarmunora commented on code/z_adventurezones/safehouse.dm:
```diff
@@ -0,0 +1,708 @@
+
+//DECORATIVE OBJECTS
+
+/obj/decal/fakeobjects/beacon
+	name = "tracking beacon"
+	desc = "A small tracking beacon. What's it doing all the way out here?"
+	icon = 'icons/obj/items/device.dmi'
+	icon_state = "beacon"
+	anchored = 1
```
maybe some kind of busted beacon, so it's clear why you can't set a teleporter to it?

MetricDuck replied:
Object description & sprite amended to better set player expectations.

Sovexe reviewed (changes requested):
whats up the icon duplication in this dmi? Shouldn't be including pre-existing sprites in a second dmi file

MetricDuck commented:
> whats up the icon duplication in this dmi? Shouldn't be including pre-existing sprites in a second dmi file

Apologies, oversight - spent all my time rechecking the dm & dmm and never thought to clean-up the dmi!

- Removed guardbuddies, unfinished feature. 
- Removed standing person, this was a mannequin for the body sprites / portrait / bust and doesn't appear in game.
- Removed gene computer - it was only different in the top two rows, so made the adjacent cabinets 32x48 and shifted the pixels across. GC now uses standard icon.
- Stripped out all idle state icons that would never be shown in-game.
- Stripped out most unused perspectives from sprites that were unlikely to be used elsewhere (these were used in previous layouts but are now redundant).

The following unused/duplicate sprites have been retained:
- adoorsingle1. Already exists elsewhere (by definition!) but it's the idle icon for the door animation. 
- Alternate perspectives for adoor - I hope to reuse these later.
- Electrical conduit copy. The existing sprite has black edges that do not pair well with the detail on the tile below (too similar), so the copy has dark grey edges instead. If there's a good way to change the colour without adding a new sprite please let me know! I experimented with the ingame colour thing cables use but it only changed the inner colour. 

Other:
- The cabinets have been retained as even if I carefully positioned existing icons together I'd still need a new sprite for the wires.
- The cloner & reclaimer were retained as they were a recolour, if this is not different enough to justify retention please let me know & I will tweak them more as I would like them to look at least slightly different to the existing and share a palette. They've already been semi-optimised via directly adding the state overlays.

MetricDuck commented:
Awaiting final self-checks in morning

Tarmunora commented on assets/maps/prefabs/prefab_safehouse.dmm:
```diff
@@ -0,0 +1,164 @@
+"oY" = (/obj/item/storage/wall/medical{dir = 1; pixel_y = 32},/turf/simulated/floor/bluewhite{dir = 1},/area/prefab/safehouse)
+"pl" = (/obj/decal/cleanable/dirt,/turf/simulated/floor/carpet/red/standard/narrow/west{dir = 1},/area/prefab/safehouse)
+"pA" = (/obj/table/auto,/obj/item/storage/toolbox/mechanical,/turf/simulated/floor/plating,/area/prefab/safehouse)
+"pJ" = (/obj/machinery/chem_dispenser,/obj/decal/tile_edge/stripe/extra_big,/turf/simulated/floor/black,/area/prefab/safehouse)
```
Please replace with a "busted dispenser" fakeobj

MetricDuck replied:
Required for puzzle, will have a think about alternatives

MetricDuck commented:
Sorry, made two last minute changes prompted by the deadline looming: unsimmed some walls I didn't want deconstructed & tweaked the xy on the handscanner objects.

MetricDuck commented:
Sorry, unsure if .dmi file is an open point so re-requested review.
